### PR TITLE
Support constructor args for forge run

### PR DESCRIPTION
**Pull request by december1981**
## Motivation
I couldn't run "forge run" for contracts which took at least one constructor argument.

## Solution

I added new constructor arg options to this cli mode which are appended to the bytecode as constructor arguments according to the abi spec.

### CODE REVIEW

**Comment by mattsse** [comment]
@brockelmore we probably want this, right?

**Comment by onbjerg** [comment]
Not sure `--constructor-args-path` makes sense, seems niche + we don't support it anywhere else?

**Comment by december1981** [comment]
> Not sure `--constructor-args-path` makes sense, seems niche + we don't support it anywhere else?

I wasn't making more work for myself :) Look at the cli for forge create, I took this additional argument from there. Presumably if someone is managing their constructor arguments from a file during create/deploy they'll want to use the same file for consistency if they're running it prior to test stuff?

File: cli/src/cmd/forge/run.rs [hunk 1]
```diff
@@ -45,6 +45,24 @@ pub struct RunArgs {
     /// Arguments to pass to the script function.
     pub args: Vec<String>,
 
+    /// constructor args calldata arguments
+    #[clap(
+        long,
+        multiple_values = true,
```
Comment (onbjerg) [inline cli/src/cmd/forge/run.rs:51]: Not sure if we are going with this anymore cc @mattsse (#1198)

**Review by onbjerg — APPROVED** [review]
otherwise lgtm

**Comment by december1981** [comment]
As a side note - I note forge run recompiles files every time it is run, even if the source hasn't changed since the last run. Is there a reason for this, or something to look into?

**Comment by gakonst** [comment]
@december1981 yeah this feels like a bug - if you have any leads, please let us know!

File: cli/src/cmd/forge/run.rs (hunk 1 above)
Comment (gakonst) [reply cli/src/cmd/forge/run.rs:51]: Let's consider this in that PR. Might be best to have a Clap constructor args struct which handles all of that logic?

**Comment by dmfxyz** [comment]
> @december1981 yeah this feels like a bug - if you have any leads, please let us know!

`run` uses `compile_files` which does not check for file changes it looks like. Same with `supress_compile`, which I noticed as part of #1256. If this isn't intentional we can fix. https://github.com/gakonst/foundry/blob/master/cli/src/compile.rs#L163

**Comment by onbjerg** [comment]
I'm not sure it's a bug. There is some information we need in `forge run` that is not always present in the artifacts (like source maps)

**Comment by brockelmore** [comment]
> @december1981 yeah this feels like a bug - if you have any leads, please let us know!

It may not be necessary anymore since we by default emit more info into the artifacts now. I would have to check

**Comment by gakonst** [comment]
@brockelmore is this PR needed or should we close?

**Comment by tynes** [comment]
> Not sure `--constructor-args-path` makes sense, seems niche + we don't support it anywhere else?

This was originally added because its possible to have constructor args larger than the max size of what is possible to be passed in via CLI. It also lets you easily commit into a git repo how you deployed a contract

**Comment by december1981** [comment]
> > Not sure `--constructor-args-path` makes sense, seems niche + we don't support it anywhere else?
> 
> This was originally added because its possible to have constructor args larger than the max size of what is possible to be passed in via CLI. It also lets you easily commit into a git repo how you deployed a contract

I noticed another issue with constructor-args-path - this line

```rust
file.split(' ').map(|s| s.to_string()).collect::<Vec<String>>()
```

It works better as?

```rust
file.split_whitespace().map(|s| s.to_string()).collect::<Vec<String>>()
```

Otherwise, if your file ends in a new line (not uncommon), you non-intuitively end up passing that newline character as an additional byte to any string argument. 

(This actually caught me out when I tried to validate a contract with one argument I had created - The abi encode of the argument in my verify script was not consistent with the published contract - because I had unintentionally published the contract using forge create with a stray newline last byte on that argument).

This also means you can put the arguments one after another on newlines in the file, and multiple contiguous whitespace does not lead to additional empty string arguments (https://doc.rust-lang.org/src/core/str/mod.rs.html#1301), which could also trip someone publishing arguments.

I can add this to this pull request (both for run.rs and create.rs)?

(Perhaps a better, more general argument format which could encode newlines, spaces and so on explicitly would be to use a json format for the arguments file -  ["arg1", "arg2", ...], but that would be a breaking change unless you interpreted the file differently depending on its extension.)

**Comment by december1981** [comment]
This is what the change in the code would like for using a json file for the arguments (where json read is only attempted if the file explicitly uses a json extension):
https://github.com/december1981/foundry/commit/e9e369a4fbde4e49f9ee096b6607acf3182911d0

**Comment by onbjerg** [comment]
@december1981 Good observation for the `--constructor-args-path` flag. Would you mind opening up an issue for it, though, since it's not necessarily blocking for this PR and is also present in other parts of Forge?

**Comment by onbjerg** [comment]
I have a general question for this PR:

There are generally two use cases for `forge run`:

- Debugging tests, which do not support constructor args
- Running scripts, which *could* support constructor args (this PR), but supports `setUp` and arguments in the script function

Since the second use case is most likely where you would need this, is there any reason you would not just have these args in the function itself, seeing as you are not able to run multiple functions in one go?

**Comment by december1981** [comment]
> @december1981 Good observation for the `--constructor-args-path` flag. Would you mind opening up an issue for it, though, since it's not necessarily blocking for this PR and is also present in other parts of Forge?

https://github.com/foundry-rs/foundry/issues/1384

**Comment by gakonst** [comment]
closing, ref https://github.com/foundry-rs/foundry/issues/2125#issuecomment-1169094418

**Pull request closed without merging**